### ISSUE TITLE
Suggestions about possible future release

### ISSUE THREAD

**Issue by Tetoq** [alexgleason/harp-gui#36]
Here a mockups about possible future improvements 
[ideas-harp.pdf](https://github.com/alexgleason/harp-gui/files/50961/ideas-harp.pdf)

**Comment by alexgleason** [comment]
I really like the improved look.

But I'm skeptical about managing the projects in that way. I think it makes it too complicated. I like the idea of having as little configuration/management as possible. It's in-line with the Harp philosophy.

**Comment by alexgleason** [comment]
This has a possible solution for #32 

I like the idea of a `File > Recent` menu better because the only configuration it needs to store is a simple text document with the file paths. And if it gets deleted, it won't have a big impact on the user.

Although now I'm even questioning that. Personally I'd much rather have a workflow where you add your common projects as favorites in your file manager. Then you can right click them and to `Open with > Harp`

**Comment by Emkaytoo** [comment]
I think a slight modification to the design @tetoq posted would provide a slightly more scannable UI for recent projects.

I just sketched them out for conversation purposes.

**First thought:**

![recent-1-sfw](https://cloud.githubusercontent.com/assets/6294225/11569398/423d4bb6-99b7-11e5-9a28-d99659c62c5f.jpg)


The bar at the bottom still opens, but then shows icons with sine waves indicating what's running and X's to show what's not. Below each icon is a toggle switch for on/off, and a folder icon for the directory. An X also closes the **Recent** menu.

**Second thought:**

![recent-2](https://cloud.githubusercontent.com/assets/6294225/11569316/e1919740-99b6-11e5-9022-248086a63d5d.jpg)

Instead of sine curves for the indicator, use the port number to indicate run status.

**Other thoughts:**
Use each icon as a link to open running projects in the user's default browser. An icon or indicator would be useful on, say, hover, but if you put the port number below the toggle switch, the "link" icon needn't be a hover.

**Comment by alexgleason** [comment]
Thanks! Nice ideas. I liked how you used the port number to represent that a server is on.

Although I'm still skeptical that a "recent projects" concept even has a place in this app. Consider what would happen in this scenario:

1. Create a new project called /my-stuff/cool-site
2. Add cool-site to Harp GUI
3.Rename cool-site to special-site
4. Reopen Harp GUI

Now Harp GUI has a managed project called cool-site, and when you try to start it, it fails.

In order to get around this, we would have to synchronize Harp GUI with the files on your system. But that's not fool proof. IMO, the file manager should be what manages your projects, not Harp GUI. If you have commonly used websites, put them in a folder called "harp apps" on your desktop so you can get to them faster.

**Comment by alexgleason** [comment]
That said, I think the "pull up drawer" design could be used to implement settings. #10

**Comment by Emkaytoo** [comment]
There are ways around the problem you described, though deciding to implement them is another thing altogether.

1. Open the GUI and check for files, failures don't load and show a fail status. A user could click that recent project and then relocate the files manually to restart, but in that case they could just re-drag to the launch screen.
* Open the GUI and check for files, only successes appear in the recents. It would be understood that recent projects that are renamed are "new" from the app's perspective.

Further, "recent projects" could be redefined as "current projects" and allow for use in multi-project environments. A multi-project manager would be useful in a team environment, where the Harp interface is used for things like review and QA by different teams/team members. Harp's [`multihost` functionality](http://harpjs.com/docs/environment/multihost) is built for just that, and I can see a use case for teams having it installed on a staging server locally and simulating production environments, or for testing on actual devices instead of through virtual machines, without having to configure each machine.

**Comment by Tetoq** [comment]
If we use the "recent project" in the menubar, like all other software, it could be just a different type of how we show the projects that the users are working on.

![harp-future_release_recents_v02](https://cloud.githubusercontent.com/assets/14060684/11577015/5487cd86-99e0-11e5-90c6-064c29482f2a.jpg)
![harp-future_release_v02](https://cloud.githubusercontent.com/assets/14060684/11577016/548aeb74-99e0-11e5-965e-5a4b41548b0d.jpg)
![harp-future_release_viewblocks_v02](https://cloud.githubusercontent.com/assets/14060684/11577014/5487794e-99e0-11e5-82d7-1156bb1f58fb.jpg)

**Comment by alexgleason** [comment]
@sintaxi @kennethormandy Any thoughts from you two?

Also pinging @JayHoltslander since he made the initial suggestion.

**Comment by Emkaytoo** [comment]
Now I'm curious: if you have the multi-project screen, why have the single-project screen at all? There's no new information presented from what I can tell. If there are no projects, then the **add** screen (the current main screen) can be the default, but once a user has projects in it the single-project screen becomes more or less irrelevant. 

Conversely, if there is a single-project screen, what additional information would be useful to have there?

**Comment by Tetoq** [comment]
In my opinion the single screen allow us to add new features in the future. For now the only possible action that we could add to the single-project is "compile App".

**Comment by sintaxi** [comment]
Wow, some really great ideas here. Here are my thoughts...

- supporting multiple projects makes a lot of sense. 
- having an obvious visual indicator for if the server is on/off is important. The toggle seems to work best.
- I would avoid using `harp multihost` because it expects that all projects are in the same directory.
- If someone turns harp-gui back on and the recent project is missing I would just let it disappear from the list. I don't see a need to show some sort of "missing project" message.
- there may be the need to remove a project form the list
- may also want to keep some sort of "publish" button in mind. there are some things are in the works ;)
- the port along does not communicate enough. better to user `http://localhost:9001` than just `9001`
- a vertical list may be easier to work with if we need to display long file paths.

**Comment by alexgleason** [comment]
Thanks for the feedback!

I definitely also agree that supporting multiple projects (#16) makes sense.

Okay, I'm looking at this from a different perspective now. Compare it to a web browser with tabs. The tabs are saved when you close the browser, then restored when you open it. But you can easily remove the tabs at any time, including a "close all tabs" option.

If a website disappears, the tab still opens and says 404. The same could be done if an app folder has moved.

I think this is a system that would work. What do you all think of a tabbed UI? @Tetoq?

**Comment by Tetoq** [comment]
I agree, is the right way to show the project. But I think the @Emkaytoo idea to have a sort of "Mission Control" is very helpful to have a entire view. Probably it could be a also a tab.

**Comment by alexgleason** [comment]
Will our users ever need to enable 5 harp sites at once? I can't envision a "mission control UI" being necessary for local development.

Check out a few screenshots from my beloved terminal application. This is the sort of UI I envision working.

![screenshot from 2015-12-05 14 18 59](https://cloud.githubusercontent.com/assets/3639540/11609999/a008395e-9b5b-11e5-84a5-fd01b865c0f9.png)
A tabbed interface. If there's a tab open, the server is on. Simple.

Then, I'll close a few tabs, and click the "history" button in the tab bar.
![screenshot from 2015-12-05 14 19 40](https://cloud.githubusercontent.com/assets/3639540/11609998/a00435f2-9b5b-11e5-91bc-f58f0b7ca247.png)
As you can see, it saved a history of the last few tabs I closed so I can easily reopen them.

Also, if I close the terminal then open it again, it restores my tabs that were previously open.

I feel like this is better due to the constantly moving nature of local development. I don't like mission control due to its sense of permanence. Generally you don't want that project from 5 weeks back to still be cluttering up your view. A tabbed UI makes it easier to manage and close active projects.

**Comment by Emkaytoo** [comment]
I can make a few cases for why Mission Control is worth it, some of them even for local development:

1. As I mentioned earlier, in team environments where they use it as a quick way to deploy to a dev/test server for internal review or testing
* In a highly-service-oriented environment, where quick changes are synced and deployed rapidly and interatively (think internal customers such as managed intranets & wikis that are built on it)
* In highly-production-oriented environments when—for better or for worse—a developer is managing concurrent projects. Having a quick-scan way of launching and reviewing would likely be beneficial
* *Gasp!* In environments where developers simply *prefer* to use graphic user interfaces over the command line, and need/would like to visually reference rendered code. Say, for exploration, iteration, reusability, or even reinforcement of an ideal from a previous project (repetition and learning).
* **This one's a little long:** Let's say there's a team where developers are pulled in too many directions (perhaps in items two and/or three above), so a non-technical member of team wants to help. This person knows that his/her team needs assistance, and there are 4 projects being developed concurrently. This person is skilled enough to know how to access a local server, but not comfortable deploying via command line. However, s/he knows that with this tool, s/he can take something off of the developers to-do list in order to allow his/her team members to focus on developing the projects instead of frequently deploying to staging servers.   <br><br>
Perhaps this team member is creating a barrier around his/her team, while simulataneously serving multiple external stakeholder (clients, account managers, etc.). Using Harp to deploy to stage from uncompiled/unprocessed code directly could now be a task for this person—say a project manager, producer, or ScrumMaster—who can pull the latest commit from their version control system and deploy without having to process. This person may be managing these 4 concurrent projects all at once, and a mission-control-style scannable UI can help this person with the frequent requests/status updates that are coming his/her way from stakeholders. As organizations/clients are comprised of humans and inherently imperfect, this is a highly likely scenario for a non-technical team member wishing to save ~~my~~ his/her team from monotony.

If you're developing a UI only for developers, and only for local, single-project development environments, then mission control may not make sense. I think we all agree the multiple project support is a good thing, and a mission control view can help the team manage the "forest", while the local development view is managing the "trees."

**Comment by Tetoq** [comment]
I fully agree on what Ian said. We have to try to think about the different types of users.
Here a first suggestion of tabs and mission control.

![harp-future_release_mcontrol](https://cloud.githubusercontent.com/assets/14060684/11631911/04ab2c8a-9cca-11e5-87b0-830477244ed0.jpg)

![harp-future_release_v03](https://cloud.githubusercontent.com/assets/14060684/11631910/049b59b8-9cca-11e5-8af3-abd125b52baf.jpg)

**Comment by isaacrg** [comment]
@Tetoq This looks awesome! This layout would be incredibly useful in helping me make project far faster.

**Comment by alexgleason** [comment]
@Emkaytoo I think you're confusing the purpose of this. I've responded below. But mission control wouldn't be seen by multiple people on a team. It would just be a way to see your own project folders on your computer. That's why I think tabs make more sense.

Also, Harp is just a server, you don't deploy _with_ it. You can deploy _it_.

> As I mentioned earlier, in team environments where they use it as a quick way to deploy to a dev/test server for internal review or testing

I don't think this is a good enough reason. Not every team will be linked up to the same internal network. Torchbox for instance has two offices as well as remote members, so this feature wouldn't benefit them. I'd rather have a feature that covers 90% of use cases than 50%. An alternative would be some form of integration with [surge.sh](https://surge.sh/) (a Harp sibling project) that allows you to deploy directly to a subdomain on the internet with a single click. And in that case they would just always be on so there's no need to manage them.

> In a highly-service-oriented environment, where quick changes are synced and deployed rapidly and interatively (think internal customers such as managed intranets & wikis that are built on it)

I agree that mission control would be great for managing deployed services, but you wouldn't run them off of your desktop computer. I suppose if your internal server had a desktop you could remote into (such as the Mac Mini at Pennebaker) this could make sense for that. But it would require additional features to make it a good user experience and I'm not sure that's the right direction to take this app in. I'd rather keep it focused on local development. Besides, most intranet servers are running headless Linux. There is no GUI, only a shell, and you would manage Harp sites using a shell.

> In highly-production-oriented environments when—for better or for worse—a developer is managing concurrent projects. Having a quick-scan way of launching and reviewing would likely be beneficial

Tabs would work for this too, right? Surely you're not going to have more than 5 or so concurrent projects.

> Gasp! In environments where developers simply prefer to use graphic user interfaces over the command line, and need/would like to visually reference rendered code. Say, for exploration, iteration, reusability, or even reinforcement of an ideal from a previous project (repetition and learning).

This is exactly what Harp GUI already does.

> This one's a little long: Let's say there's a team where developers are pulled in too many directions (perhaps in items two and/or three above), so a non-technical member of team wants to help. This person knows that his/her team needs assistance, and there are 4 projects being developed concurrently. This person is skilled enough to know how to access a local server, but not comfortable deploying via command line. However, s/he knows that with this tool, s/he can take something off of the developers to-do list in order to allow his/her team members to focus on developing the projects instead of frequently deploying to staging servers.

Well, you know a mission control wouldn't be synced across multiple devices. You wouldn't be able to see what Teto's running, for instance. It's only for managing your own stuff.

> If you're developing a UI only for developers, and only for local, single-project development environments, then mission control may not make sense. I think we all agree the multiple project support is a good thing, and a mission control view can help the team manage the "forest", while the local development view is managing the "trees."

It is meant for developers, but it should be easy enough for anyone to use it. But I think you're misunderstanding what we mean by mission control here. I think you're asking for a new feature that lets multiple people review projects on a team over the internet.

**Comment by Emkaytoo** [comment]
@alexgleason I hear what you're saying. Below are some clarifications:

I'm not suggesting a team view. I'm suggesting that one person could be responsible for managing multiple sites with it, and that the "local machine" could be a "local server" that itself has a UI. I wouldn't need to see what each developer is running locally, but in, say, a scrum environment when a PO is accepting completed work items, those would need to be put somewhere for acceptance before being show in a sprint review with stakeholders. 

I'll start a new thread where we can discuss "team" features and other PM attributes.

**Comment by alexgleason** [comment]
Awesome. Let's continue discussing mission control in #16
